### ISSUE TITLE
Service Accounts - delayed deployments / failure logs - Unauthenticated

### ISSUE THREAD

**Issue by bryfry** [alta3/kubernetes-the-alta3-way#21]
Many instances documented below, TL;DR in-cluster service accounts fail until they don't

### Tested hypothesises: 
 - Etcd lacks write speed (fsync) to keep up. 
   - Increasing CPU/Ram did not elevate this issue
   - Running an entire cluster from ramdisk did not elevate this issue
 - Service accounts are being provisioned incorrectly
   - Forcing authentication using issued bearer tokens did not fail 
   - All pod failures eventually succeed after redeployment

**Comment by bryfry** [comment]
# Calico service accounts:

After the pod gets restarted the error goes away, but the number of restarts seems like a roll of the dice before the pod lands on a successful startup. Debug logging on, same opaque details about the connection to the apiserver.

```
2022-03-08 21:30:07.677 [INFO][9] startup/startup.go 425: Early log level set to debug
....
2022-03-08 21:30:07.738 [ERROR][9] startup/startup.go 180: Failed to read Node from datastore error=Unauthorized
```

Full log:

```
2022-03-08 21:30:07.677 [INFO][9] startup/startup.go 425: Early log level set to debug
2022-03-08 21:30:07.677 [INFO][9] startup/utils.go 129: Using stored node name node-1 from /var/lib/calico/nodename
2022-03-08 21:30:07.677 [INFO][9] startup/utils.go 139: Determined node name: node-1
2022-03-08 21:30:07.677 [INFO][9] startup/startup.go 94: Starting node node-1 with version v3.22.1
2022-03-08 21:30:07.678 [DEBUG][9] startup/load.go 95: Datastore type isn't set, trying to detect it
2022-03-08 21:30:07.678 [DEBUG][9] startup/load.go 97: EtcdEndpoints specified, detected etcdv3.
2022-03-08 21:30:07.678 [DEBUG][9] startup/client.go 30: Using datastore type 'etcdv3'
2022-03-08 21:30:07.679 [INFO][9] startup/startup.go 430: Checking datastore connection
2022-03-08 21:30:07.679 [DEBUG][9] startup/etcdv3.go 357: Processing Get request model-etcdKey=Node(foo) rev=""
2022-03-08 21:30:07.679 [DEBUG][9] startup/etcdv3.go 381: Calling Get on etcdv3 client etcdv3-etcdKey="/calico/resources/v3/projectcalico.org/nodes/foo" model-etcdKey=Node(foo) rev=""
2022-03-08 21:30:07.705 [DEBUG][9] startup/etcdv3.go 388: No results returned from etcdv3 client etcdv3-etcdKey="/calico/resources/v3/projectcalico.org/nodes/foo" model-etcdKey=Node(foo) rev=""
2022-03-08 21:30:07.705 [INFO][9] startup/startup.go 454: Datastore connection verified
2022-03-08 21:30:07.705 [INFO][9] startup/startup.go 104: Datastore is ready
2022-03-08 21:30:07.705 [DEBUG][9] startup/etcdv3.go 357: Processing Get request model-etcdKey=Node(node-1) rev=""
2022-03-08 21:30:07.705 [DEBUG][9] startup/etcdv3.go 381: Calling Get on etcdv3 client etcdv3-etcdKey="/calico/resources/v3/projectcalico.org/nodes/node-1" model-etcdKey=Node(node-1) rev=""
2022-03-08 21:30:07.733 [INFO][9] startup/startup.go 156: Unauthorized to query kubeadm configmap, assuming not on kubeadm. CIDR detection will not occur. error=Unauthorized
2022-03-08 21:30:07.735 [INFO][9] startup/startup.go 171: Unauthorized to query rancher configmap, assuming not on rancher. CIDR detection will not occur. error=Unauthorized
2022-03-08 21:30:07.738 [ERROR][9] startup/startup.go 180: Failed to read Node from datastore error=Unauthorized
2022-03-08 21:30:07.738 [WARNING][9] startup/utils.go 49: Terminating
```

Comparing to a successful launch they diverge on:

```
2022-03-08 21:30:06.032 [DEBUG][9] startup/etcdv3.go 381: Calling Get on etcdv3 client etcdv3-etcdKey="/calico/resources/v3/projectcalico.org/nodes/node-2" model-etcdKey=Node(node-2) rev=""
```

On the success side:
```
2022-03-08 21:30:07.073 [DEBUG][9] startup/reachaddr.go 28: Auto-detecting IPv4 CIDR by reaching destination 8.8.8.8
2022-03-08 21:30:07.074 [INFO][9] startup/reachaddr.go 47: Auto-detected address by connecting to remote Destination="8.8.8.8" IP=10.13.76.227
<--snip-->
```

On the failed invocation:
```
2022-03-08 21:30:06.051 [INFO][9] startup/startup.go 156: Unauthorized to query kubeadm configmap, assuming not on kubeadm. CIDR detection will not occur. error=Unauthorized
2022-03-08 21:30:06.053 [INFO][9] startup/startup.go 171: Unauthorized to query rancher configmap, assuming not on rancher. CIDR detection will not occur. error=Unauthorized
2022-03-08 21:30:06.055 [ERROR][9] startup/startup.go 180: Failed to read Node from datastore error=Unauthorized
2022-03-08 21:30:06.055 [WARNING][9] startup/utils.go 49: Terminating
```

**Comment by bryfry** [comment]
# CoreDNS

CoreDNS k8s plugin failures are also related.  Re-deployment of the coredns pod (not node dependent) and sometimes it comes up and works and other times it fails.

**Always Working:**
 - forward resolution

**Intermittent failure:**
 - Kubernetes resolution of pods

**Failure mode:**

```
[WARNING] plugin/kubernetes: starting server with unsynced Kubernetes API
```

If this log happens I know coredns will never recover and will never resolve any in-cluster addresses.  Changing nothing -- only deleting the coredns pod and allowing the deployment to reprovision the pod -- on startup the service seems to be able to sync with kubernetes api about 50% of the time.

**Comment by bryfry** [comment]
# Containerd / Kubelet network sandbox

### fluentd logger pod

logger pod
```
Events:
  Type     Reason                  Age                From               Message
  ----     ------                  ----               ----               -------
  Normal   Scheduled               25m                default-scheduler  Successfully assigned default/logger to node-3
  Warning  FailedCreatePodSandBox  25m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "51b8dd9b5b29213de994462782e66dc8e077a16d5ba09d7b3b38cb298858c9e8": Unauthorized
  Warning  FailedCreatePodSandBox  25m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "31e49b23c0e0cc8fae2db290f59564b1a1df382d8515db0b71f899a8a4f823c0": Unauthorized
  Warning  FailedCreatePodSandBox  24m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "3d8a1b429ceb0d062ab9625216452b4f3c2865d5eccaab274fb43d0c5abe8d2f": Unauthorized
  Warning  FailedCreatePodSandBox  24m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "afe03a59a5752de22b4ef47cc13207ae3bedd3bd5c9cd296a04fa5d11190ea11": Unauthorized
  Warning  FailedCreatePodSandBox  24m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "5818316e4187e91f0994faae9a707ebe7a55632d7824beaddad96223ea13f08b": Unauthorized
  Warning  FailedCreatePodSandBox  24m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "30460e18b8068ed4795ea91241ff40032f9a091c09be397bd83f6bb138fd43a2": Unauthorized
  Warning  FailedCreatePodSandBox  24m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "68e97f0bd351c3ab7615e4cee0ed66b2f65d262bf1f35f773996e44d84783bc3": Unauthorized
  Warning  FailedCreatePodSandBox  23m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "ffd1b5e9c1af5c0ae33e860de7cb7b2d67cd0f52227557c7f25dc5ebe7b3fdbe": Unauthorized
  Warning  FailedCreatePodSandBox  23m                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "8b25d1f9e3bfe3103192b5a84a6157960393d0ef637c55fad974de5660695500": Unauthorized
  Warning  FailedCreatePodSandBox  22m (x6 over 23m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "8c99769690d061a3b25057e1ab859d0bc4e9a0558fd2cd9ab52e549cccdc7bdb": Unauthorized
  Normal   Pulling                 21m                kubelet            Pulling image "busybox"
  Normal   Pulled                  21m                kubelet            Successfully pulled image "busybox" in 364.378637ms
  Normal   Created                 21m                kubelet            Created container count
  Normal   Started                 21m                kubelet            Started container count
  Normal   Pulling                 21m                kubelet            Pulling image "k8s.gcr.io/fluentd-gcp:1.30"
  Normal   Pulled                  21m                kubelet            Successfully pulled image "k8s.gcr.io/fluentd-gcp:1.30" in 14.83659972s
  Normal   Created                 21m                kubelet            Created container count-agent
  Normal   Started                 21m                kubelet            Started container count-agent
```

http-logger pod
```
Events:
  Type     Reason                  Age    From               Message
  ----     ------                  ----   ----               -------
  Normal   Scheduled               7m39s  default-scheduler  Successfully assigned default/http-logger to node-2
  Warning  FailedCreatePodSandBox  7m37s  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "8fe609dab69676ba5b3a156636e23cd83a2aff9ddeb8c2043e05a15c93f04b77": Unauthorized
  Warning  FailedCreatePodSandBox  7m26s  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "be963e0713d4cd0edfec8e36a146526dcef8d150850a2064a2928a895acdf33a": Unauthorized
  Warning  FailedCreatePodSandBox  7m11s  kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "0f5296297e2285500cea05c471d07d3ba8a8714c9d7c9e6589d68a44758876fd": Unauthorized
  Normal   Pulling                 6m59s  kubelet            Pulling image "busybox"
  Normal   Pulled                  6m59s  kubelet            Successfully pulled image "busybox" in 650.958395ms
  Normal   Created                 6m58s  kubelet            Created container count
  Normal   Started                 6m58s  kubelet            Started container count
  Normal   Pulling                 6m58s  kubelet            Pulling image "fluentd"
  Normal   Pulled                  6m55s  kubelet            Successfully pulled image "fluentd" in 3.211615487s
  Normal   Created                 6m55s  kubelet            Created container count-agent
  Normal   Started                 6m55s  kubelet            Started container count-agent
```

**Comment by bryfry** [comment]
# Raw curl with service account token

Testing a specific service account yields a different response for multiple requests generated in a row.  In this case the expected return was the 403. 401 indicates that the bearer token failed. 

```
root@authpod:/# curl --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt https://kubernetes.default/api/v1/amespaces/default/pods -H "Authorization: Bearer $TOKEN"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "pods is forbidden: User \"system:serviceaccount:default:pod-user\" cannot list resource \"pods\" in API group \"\" in the namespace \"default\"",
  "reason": "Forbidden",
  "details": {
    "kind": "pods"
  },
  "code": 403
}root@authpod:/# curl --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt https://kubernetes.default/api/v1/amespaces/default/pods -H "Authorization: Bearer $TOKEN"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
```

**Comment by bryfry** [comment]
# Emergent behavior of kube-apiserver and etcd

Watching the logs of all three kube-apiservers where kube-proxy load balanced the kube-apiservice to each of the controllers we were able to observe:
 
- Only one kube-apiserver would return the successfully authenticated session
- No changes to etcd (moving leader between systems, turning some off, etc) changed the behavior

This lends itself to demonstrating that etcd data replication is not the root of the issue. A configuration issue in kube-apiserver is the likely problem.

**Comment by bryfry** [comment]
# `--service-account-issuer strings`

https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/

> Identifier of the service account token issuer. The issuer will assert this identifier in "iss" claim of issued tokens. This value is a string or URI. If this option is not a valid URI per the OpenID Discovery 1.0 spec, the ServiceAccountIssuerDiscovery feature will remain disabled, even if the feature gate is set to true. It is highly recommended that this value comply with the OpenID spec: https://openid.net/specs/openid-connect-discovery-1_0.html. In practice, this means that service-account-issuer must be an https URL. It is also highly recommended that this URL be capable of serving OpenID discovery documents at {service-account-issuer}/.well-known/openid-configuration. When this flag is specified multiple times, the first is used to generate tokens and all are used to determine which issuers are accepted.

**TL;DR If you don't set this flag for each of your kube-apiservers they will not accept service account tokens issued by other apiservers**

**Comment by bryfry** [comment]
Here's the fix

https://github.com/alta3/kubernetes-the-alta3-way/blob/c3463f4643e8e49d2f0887d854ceaf420a8eba01/roles/controller_install/templates/kube-apiserver.service.j2#L30-L32
